pkg/gadgets/trace/tcpretrans/tracer/bpf/tcpretrans.bpf.c

```diff
@@ -0,0 +1,155 @@
+
+#include <bpf/bpf_helpers.h>
+#include <bpf/bpf_core_read.h>
+#include <bpf/bpf_tracing.h>
+#include <bpf/bpf_endian.h>
+
+#define GADGET_TYPE_TRACING
+#include <sockets-map.h>
+
+#include "tcpretrans.h"
+
+/* Define here, because there are conflicts with include files */
+#define AF_INET		2
+#define AF_INET6	10
+
+// we need this to make sure the compiler doesn't remove our struct
+const struct event *unusedevent __attribute__((unused));
+
+struct {
+	__uint(type, BPF_MAP_TYPE_PERF_EVENT_ARRAY);
+	__uint(key_size, sizeof(__u32));
+	__uint(value_size, sizeof(__u32));
+} events SEC(".maps");
+
+// This struct is defined according to the following format file:
+// /sys/kernel/tracing/events/tcp/tcp_retransmit_skb/format
+struct tcp_retransmit_skb_ctx {
+	/* The first 8 bytes are not allowed to read */
+	unsigned long pad;
+
+	void *skbaddr;
+	void *skaddr;
+	int state;
+	__u16 sport;
+	__u16 dport;
+	__u8 saddr[4];
+	__u8 daddr[4];
+	__u8 saddr_v6[16];
+	__u8 daddr_v6[16];
+};
+
+// This struct is the same as struct tcphdr in vmlinux.h but with flags defined as single field instead of bitfield
+struct tcphdr_with_flags {
+	__be16 source;
+	__be16 dest;
+	__be32 seq;
+	__be32 ack_seq;
+	__u16 res1: 4;
+	__u16 doff: 4;
+	__u8 flags;
+	__be16 window;
+	__sum16 check;
+	__be16 urg_ptr;
+};
+
+static __always_inline int __trace_tcp_retrans(void *ctx, struct sock *sk, struct sk_buff *skb)
+{
+	if (sk == NULL)
+		return 0;
+
+	if (BPF_CORE_READ_BITFIELD_PROBED(sk, sk_protocol) != IPPROTO_TCP)
+		return 0;
+
+	struct inet_sock *sockp = (struct inet_sock *)sk;
+	struct task_struct *task = (struct task_struct*) bpf_get_current_task();
+	__u64 pid_tgid = bpf_get_current_pid_tgid();
+
+	struct event event = {};
+	event.timestamp = bpf_ktime_get_boot_ns();
+	event.af = BPF_CORE_READ(sk, __sk_common.skc_family);
+	event.state = BPF_CORE_READ(sk, __sk_common.skc_state);
+
+	bpf_get_current_comm(&event.proc_current.task, sizeof(event.proc_current.task));
+	event.proc_current.pid = pid_tgid >> 32;
+	event.proc_current.tid = (__u32)pid_tgid;
+	event.proc_current.mount_ns_id = (u64) BPF_CORE_READ(task, nsproxy, mnt_ns, ns.inum);
+
+	// The tcp_retransmit_skb tracepoint is fired with a skb that does not
+	// contain the TCP header because the TCP header is built on a cloned skb
+	// we don't have access to.
+	// skb->transport_header is not set: skb_transport_header_was_set() == false.
+	// Instead, we have to read the TCP flags from the TCP control buffer.
+	struct tcp_skb_cb *tcb = (struct tcp_skb_cb *)&(skb->cb[0]);
+	bpf_probe_read_kernel(&event.tcpflags, sizeof(event.tcpflags), &tcb->tcp_flags);
+
+	BPF_CORE_READ_INTO(&event.dport, sk, __sk_common.skc_dport);
+	if (event.dport == 0)
+		return 0;
+
+	BPF_CORE_READ_INTO(&event.sport, sockp, inet_sport);
+	if (event.sport == 0)
+		return 0;
+
+	switch (event.af) {
+	case AF_INET:
+		BPF_CORE_READ_INTO(&event.daddr_v4, sk, __sk_common.skc_daddr);
+		if (event.daddr_v4 == 0)
+			return 0;
```

Comment:
Do we need these checks?

Reply:
I don't know. There is the same pattern in tcpdrop.bpf.c and tcptracer.bpf.c.